completions/fish/wl-paste.fish

```diff
@@ -0,0 +1,52 @@
+function __wayland_seats --description 'Print all wayland seats'
+    if type -q weston-info
+        weston-info 2>/dev/null | sed -n '/wl_seat/{n;s/\s*name: //;p}'
+    else if type -q loginctl
+        loginctl --no-legend --no-pager list-seats 2>/dev/null
+    else
+        # Fallback seat
+        echo "seat0"    
+    end
+end
+
+function __wl_paste_types --description 'Print types with context'
+    set -l type (commandline -o)
+
+    for i in (seq (count $type))
+        if [ "$type[$i]" = '--primary' ]; or [ "$type[$i]" = '-p' ]
+            set clip "primary"
+        else if [ "$type[$i]" = '--seat' ]; or [ "$type[$i]" = '-s' ]
+            set seat $type[(math $i + 1)]
+        end
+    end
+
+    # Note fish does not handle passing unset variables
+    # to commands well, thus setting clip to "--primary" and passing
+    # that to wl-paste wont work, so if statements are used instead
+    if test -n "$seat"; and test -n "$clip"
+        wl-paste --seat "$seat" -p -l
+    else if test -n "$seat"
+        wl-paste --seat "$seat" -l
+    else if test -n "$clip"
+        wl-paste -p -l
+    else
+        wl-paste -l
+    end
+end
+
+function __all_commands --description 'Generate list of all commands in path'
+    set -l pos (commandline -o)
+    complete -c wl-paste -x -a "(__fish_complete_subcommand --fcs-skip=(count $pos))"
+    complete -c wl-paste -f # Needed otherwise it offers files after __fish_complete_subcommand is called
+    ls (echo $PATH | tr ' ' '\n') | grep -v '/' | grep . | sort
+end
```

Comment:
Is this all really necessary? I'm pretty sure `__fish_complete_subcommand` handles most of the logic. Shouldn't it be enough to just pass it to `complete -c wl-paste -s w -l watch` below?

In particular, you're *really* not supposed to manually generate the list of programs by looking in `PATH`.

Reply:
Passing `__fish_complete_subcommand` will just return a list of files as completions; it needs to be called and we need to pass programs in $PATH. The `--fcs-skip=(count $pos)` logic seems to be handled within the command so I've removed it. For the list of programs there doesn't really seem to be something equivalent to bash's `compgen -c` there is `command --all` is supposed to return all all external commands, but running it returns nothing.

Reply:
Actually upon further testing `__fish_complete_subcommand` will only show the local directory contents *at first.* Once you type for example `r` it will than list executable's. The original code is probably closer to zsh's and bash's completions, but if you want to we can just rely on  `__fish_complete_subcommand`

Reply:
We should definitely rely on `__fish_complete_subcommand`, we just need to figure out how to use it properly :slightly_smiling_face:

I've spent some time staring at `__fish_complete_subcommand`'s source code and experimenting, and came up with the following:

```fish
complete -c wl-paste -f
complete -c wl-paste -f -n '__fish_contains_opt -s w watch' -a "(__fish_complete_subcommand -- -s --seat -t --type)"
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s h -l help -d 'Display a help message'
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s v -l version -d 'Display version info'
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s n -l no-newline -d 'Do not append a newline character'
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s l -l list-types -d 'Instead of pasting, list the offered types'
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s p -l primary -d 'Use the "primary" clipboard'
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s w -l watch -d 'Run a command each time the selection changes'
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s t -l type -x -d 'Override the inferred MIME type for the content' -a "(__wl_paste_types)"
complete -c wl-paste -n '__fish_not_contain_opt -s w watch' -s s -l seat -x -d 'Pick the seat to work with' -a "(__wayland_seats)"
```

It does seem to work exactly as I want it to. Please check if it also works on your machine, and if it does, incorporate it into your commit.

Here's what this does: fish seemingly insists on only letting options provide argument completions for *one* argument value, and in our case `--watch` accepts an arbitrary number of arguments (the rest of them). So we tell fish that it's `wl-paste` itself that accepts a subcommand, but only if `-w`/`--watch` has been specified previously, and that the rest of options are only valid if `-w`/`--watch` has *not* been specified previously.

Reply:
This works, but for `--watch` will offer no completions until a letter is typed; I think this presents a discoverability problem as there's no indication it actually offers completions, as well as it doesn't give a initial indication as to what is an acceptable input.

Reply:
I agree, but isn't that how command completion works everywhere else in fish? Both when completing a simple command (try completing at the start of a new line prior to typing anything) and when completing a subcommand (try `sudo <TAB>`).

You should probably file an issue at their repo to fix this.

Reply:
Yeah, fair enough it's probably better to maintain consistency with the rest of the shell.